### PR TITLE
Add a midi visualizer

### PR DESCRIPTION
**Pull request by notwaldorf**
- made `mm.Visualizer` in `core`, which makes something look like this:
screen shot 2018-08-17 at 4 58 01 pm

- in the demo, I connected it to a Player to make it play in sync. I had originally made this as a web component, but loading web components is a bit icky with `es5`, so I passed on it.

There's a `TODO` in the code to make the repainting slightly better and more optimal, but I wanted to see first if this was barking up the right tree before I went down the road of fixing that.

Also: the demo is ugly, but I am planning a followup PR to make all the demos a bit prettier (and easier to understand their input/output)

I uhhhh don't know how to test this, or if it's worth testing. 😭

### CODE REVIEW

**Comment by notwaldorf** [comment]
(though I guess I can make a /components folder, sibling to music and sketch, that has es6 web components and doesn’t get packaged with magentamusic. Wdyt?)

File: music/src/core/visualizer.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {INoteSequence, NoteSequence} from '../protobuf';
+
+/**
+ * A NoteSequence visualizer.
+ *
```
Comment (adarob) [inline music/src/core/visualizer.ts]: give a brief description of what it looks like

File: music/src/core/visualizer.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+ */
+export class Visualizer {
+  private config: VisualizerConfig;
+  private ctx: CanvasRenderingContext2D;
+  private height: number;
+  public noteSequence: INoteSequence;
+  private minPitch = 100;
+  private maxPitch = -1;
```
Comment (adarob) [inline music/src/core/visualizer.ts]: shouldn't these be set in the constructor (w/ defaults)?

File: music/src/core/visualizer.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      this.config.maxPitch += 2;
+    }
+
+    // Height of the canvas based on the range of pitches in the sequence
+    const height =
+        (this.config.maxPitch - this.config.minPitch) * this.config.noteHeight;
+
+    // Calculate a nice width based on the length of the sequence we're playing.
+    const numNotes = this.noteSequence.notes.length;
+    const endTime = this.getNoteEndTime(this.noteSequence.notes[numNotes - 1]);
```
Comment (adarob) [inline music/src/core/visualizer.ts]: We can't guarantee the notes are ordered, but there are a `total_time` and `total_quantized_steps` fields in the NoteSequence.

File: music/src/core/visualizer.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+   *   @param sequence The `NoteSequence` to be visualized
+   *   @param canvas The element where the visualization should be displayed.
+   *   @param config Visualization configuration options
+   */
+
+  constructor(
+      sequence: INoteSequence, canvas: HTMLCanvasElement,
+      config = {} as VisualizerConfig) {
+    this.config = {
+      noteHeight: config.noteHeight || 6,
```
[the remark this answers is not in the document]
Comment (notwaldorf) [reply music/src/core/visualizer.ts]: I don't know how else to do it but if we figure it out, we can change it? How to typescript.

**Pull request merged**